### ISSUE TITLE
Exceptions when a TCPSocket connection timed out in and out of the Reactor using Kernel.timeout

### ISSUE THREAD

**Issue by HoneyryderChuck** [celluloid/celluloid-io#97]
I'm fighting against the non-establishment of connections using Celluloid::IO::TCPSocket s. Currently my issue is with connections timing out. 

Let's say I have an "timeoutable" hostname. 

``` ruby
Celluloid::IO::TCPSocket.new("timeoutable", 23)
(10 seconds later...)
Errno::ETIMEDOUT: Connection timed out - connect(2)
```

Which is consistent with what happens when using a regular TCPSocket. 

Now, when I do that inside an actor...

``` ruby
Celluloid A
  include Celluloid::IO
  def establish
    timeout(5) do
      @tcp = TCPSocket.new("timeoutable", 23)
    end
  end
end
a.establish
```

It raises a Celluloid::Task::TimeoutError, which makes sense (but it kills the actor). 

Now, other case:

``` ruby
Celluloid A
  include Celluloid::IO
  def establish
    Kernel.send(:timeout, 5) do
      @tcp = TCPSocket.new("timeoutable", 23)
    end
  end
end
a.establish
```

This one raises an "execution expired" exception coming from the reactor. And kills the actor.

So, there are two issues for me. Let's say I have an actor whose task is to perform a task on x remote devices. Being that for some I cannot connect on time, how can I rescue the connection timeout and not kill the actor in the process, but simple handle the logic, mark the remote location as unreachable and move on with my actor life? The way I tested, I can rescue Errno::ETIMEDOUT and Celluloid::Task::TimeoutError on the method and keep the actor alive. But I cannot rescue the third one, because that exception is thrown after the actor has been killed, so no way to keep him alive. 

The second is regarding that last actor. As you have seen, I have used Kernel.timeout inside the reactor loop, which clearly doesn't play along very well. I know your first question is "uh, why you used Kernel.timeout instead of Actor#timeout as in the second example?", but I'm using a wrapper whom I pass a proxy. This wrapper doesn't know anything about Celluloid, hence it calls normal. timeout. i don't have any influence. As I've seen, I can only use this timeout inside the actor (an instance method, therefore). How could I access the current actor and call the #timeout method in such a case?

I hope the second question can lead to an answer of the first one :)

**Comment by ioquatix** [comment]
Is this still a problem?

**Comment by TiagoCardoso1983** [comment]
yup, at least until the timeout extensions are integrated in celluloid.
